packages/nextra-theme-docs/src/contexts/active-anchor.tsx

```diff
@@ -24,22 +27,89 @@ const SetActiveAnchorContext = createContext<
   Dispatch<SetStateAction<ActiveAnchor>>
 >(v => v)
 
+const IntersectionObserverContext = createContext<IntersectionObserver | null>(null)
+const slugs = new WeakMap()
+const SlugsContext = createContext<WeakMap<any, any>>(slugs)
 // Separate the state as 2 contexts here to avoid
 // re-renders of the content triggered by the state update.
 export const useActiveAnchor = () => useContext(ActiveAnchorContext)
 export const useSetActiveAnchor = () => useContext(SetActiveAnchorContext)
 
+export const useIntersectionObserver = () => useContext(IntersectionObserverContext)
+export const useSlugs = () => useContext(SlugsContext)
+
 export const ActiveAnchorProvider = ({
   children
 }: {
   children: ReactNode
 }): ReactElement => {
   const [activeAnchor, setActiveAnchor] = useState<ActiveAnchor>({})
+  const [observer] = useState<IntersectionObserver | null>(() => {
```

Comment:
we can move `observer` outside of provider definition to avoid confusing `const [observer] = useState` without `setState`

Reply:
It needs to use `setActiveAnchor `, so we still need to initialize it in provider. But i think i could use a `useRef` here.